File: 39. Combination Sum.cpp
class Solution {
public:
    vector<vector<int>> combinationSum(vector<int>& candidates, int target) {
        vector<vector<int> > res;
        vector<int> path;
        sort(candidates.begin(), candidates.end());
        helper(candidates, 0, target, path, res);
        return res;
    }

    void helper(vector<int>& candidates, int idx, int target, vector<int> path, vector<vector<int> > &res) {
        if (target < 0) {
            return;
        } else if (target == 0) {
            res.push_back(path);
            return;
        } else if (candidates[idx] > target) {
            return;
        }
        while (idx < candidates.size()) {
            path.push_back(candidates[idx]);
            helper(candidates, idx, target - candidates[idx], path, res);
            path.pop_back();
            idx++;
        }
    }
};

File: java/39. Combination Sum.java
class Solution {
    public List<List<Integer>> combinationSum(int[] candidates, int target) {
        List<List<Integer>> res = new ArrayList<>();
        dfs(candidates, 0, target, new ArrayList<Integer>(), res);
        return res;
    }

    private void dfs(int[] candidates, int start, int target, ArrayList<Integer> path, List<List<Integer>> res) {
        if (target == 0) {
            res.add(new ArrayList<>(path));
            return;
        }
        for (int i = start; i < candidates.length; ++i) {
            if (candidates[i] <= target) {
                path.add(candidates[i]);
                dfs(candidates, i, target - candidates[i], path, res);
                path.remove(path.size() - 1);
            }
        }
    }
}
